### PR TITLE
Add a timeout to every command (enforced on device and host)

### PR DESCRIPTION
**Pull request by goderbauer**

### CODE REVIEW

File: packages/flutter_driver/lib/src/driver.dart [hunk 1]
```diff
@@ -227,17 +227,28 @@ class FlutterDriver {
...
     } catch (error, stackTrace) {
       throw new DriverError(
         'Failed to fulfill ${command.runtimeType} due to remote error',
         error,
         stackTrace
       );
     }
+    if (response['isError']) {
+      throw new DriverError('Error in Flutter application: ${response['response']}');
+    }
```
Comment (yjbanov) [inline packages/flutter_driver/lib/src/driver.dart]: style nit: do not use curlies in single-statement if blocks.

File: packages/flutter_driver/lib/src/extension.dart [hunk 2]
```diff
@@ -74,15 +74,15 @@ class _FlutterDriverExtension {
...
-      'get_health': GetHealth.deserialize,
-      'get_render_tree': GetRenderTree.deserialize,
-      'tap': Tap.deserialize,
-      'get_text': GetText.deserialize,
-      'scroll': Scroll.deserialize,
-      'scrollIntoView': ScrollIntoView.deserialize,
-      'setInputText': SetInputText.deserialize,
-      'submitInputText': SubmitInputText.deserialize,
-      'waitFor': WaitFor.deserialize,
+      'get_health': (json) => new GetHealth.deserialize(json),
```
Comment (yjbanov) [inline packages/flutter_driver/lib/src/extension.dart]: style nit: always use types: `String json`

File: packages/flutter_driver/lib/src/extension.dart [hunk 3]
```diff
@@ -108,21 +108,34 @@ class _FlutterDriverExtension {
...
     try {
-      String commandKind = params['command'];
       CommandHandlerCallback commandHandler = _commandHandlers[commandKind];
       CommandDeserializerCallback commandDeserializer =
           _commandDeserializers[commandKind];
       if (commandHandler == null || commandDeserializer == null)
         throw 'Extension $_extensionMethod does not support command $commandKind';
       Command command = commandDeserializer(params);
-      return (await commandHandler(command)).toJson();
+      var response = await commandHandler(command).timeout(command.timeout);
```
Comment (yjbanov) [inline packages/flutter_driver/lib/src/extension.dart]: style nit: do not use `var`.

File: packages/flutter_driver/lib/src/message.dart [hunk 4]
```diff
@@ -5,11 +5,25 @@
...
+
   /// Identifies the type of the command object and of the handler.
   String get kind;
 
   /// Serializes this command to parameter name/value pairs.
-  Map<String, String> serialize();
+  Map<String, String> serialize() => {
+        'command': kind,
+        'timeout': '${timeout.inMilliseconds}',
+      };
```
Comment (yjbanov) [inline packages/flutter_driver/lib/src/message.dart]: Hmm, I'd format this like this:

```dart
Map<String, String> serialize() => {
  'command': kind,
  'timeout': '${timeout.inMilliseconds}',
};
```

Also, the map literal needs to be typed according to Flutter's style guide.

**Review by yjbanov — APPROVED** [review]
Just a bunch of style nits.

File: packages/flutter_driver/lib/src/message.dart (hunk 4 above)
Comment (goderbauer) [inline packages/flutter_driver/lib/src/message.dart]: Done

File: packages/flutter_driver/lib/src/extension.dart (hunk 3 above)
Comment (goderbauer) [inline packages/flutter_driver/lib/src/extension.dart]: Done

File: packages/flutter_driver/lib/src/extension.dart (hunk 2 above)
Comment (goderbauer) [inline packages/flutter_driver/lib/src/extension.dart]: Done

File: packages/flutter_driver/lib/src/driver.dart (hunk 1 above)
Comment (goderbauer) [inline packages/flutter_driver/lib/src/driver.dart]: Done.

**Comment by goderbauer** [comment]
Also added some test. @yjbanov PTAL

File: packages/flutter_driver/test/flutter_driver_test.dart [hunk 5]
```diff
@@ -279,6 +281,45 @@ void main() {
         expect(timeline.events.single.name, 'test event');
       });
     });
+
+    group('sendCommand error conditions', () {
+      test('local timeout', () async {
+        when(mockIsolate.invokeExtension(any, any)).thenAnswer((Invocation i) {
+          return new Future<Map<String, dynamic>>.delayed(
```
Comment (yjbanov) [inline packages/flutter_driver/test/flutter_driver_test.dart]: Presumably you want this function to never return a value. Therefore, can this be simply:

```dart
return new Completer<Map<String, dynamic>>.future;
```

?

**Review by yjbanov — APPROVED** [review]

File: packages/flutter_driver/test/flutter_driver_test.dart (hunk 5 above)
Comment (goderbauer) [inline packages/flutter_driver/test/flutter_driver_test.dart]: Yes! Updated.

**Pull request merged**